pkg/controllers/clusters/clusterlifecycle/clusterlifecycle_controller.go

```diff
@@ -0,0 +1,210 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterlifecycle
+
+import (
+	"context"
+	"reflect"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	clusterapi "github.com/clusternet/clusternet/pkg/apis/clusters/v1beta1"
+	clusternetClientSet "github.com/clusternet/clusternet/pkg/generated/clientset/versioned"
+	clusternetInformers "github.com/clusternet/clusternet/pkg/generated/informers/externalversions"
+	clusterlisters "github.com/clusternet/clusternet/pkg/generated/listers/clusters/v1beta1"
+)
+
+const (
+	ClusterMonitorPeriod             = 20 * time.Second
+	DefaultClusterMonitorGracePeriod = 9 * time.Minute
+
+	// ClusterMonitorGracePeriod should be N times more than cluster's HeartbeatFrequencySeconds
+	ClusterMonitorGracePeriodFactor = 3
+)
+
+// Controller is a controller that manages cluster's lifecycle
+type Controller struct {
+	ctx context.Context
+
+	clusternetClient clusternetClientSet.Interface
+
+	clusterLister clusterlisters.ManagedClusterLister
+	clusterSynced cache.InformerSynced
+}
+
+func NewController(ctx context.Context, clusternetClient clusternetClientSet.Interface,
+	clusternetInformerFactory clusternetInformers.SharedInformerFactory) *Controller {
+	return &Controller{
+		ctx:              ctx,
+		clusternetClient: clusternetClient,
+		clusterLister:    clusternetInformerFactory.Clusters().V1beta1().ManagedClusters().Lister(),
+		clusterSynced:    clusternetInformerFactory.Clusters().V1beta1().ManagedClusters().Informer().HasSynced,
+	}
+}
+
+func (c *Controller) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting cluster lifecycle controller")
+	defer klog.Infof("Shutting down cluster lifecycle controller")
+
+	// Wait for the caches to be synced before starting monitor lifecycle
+	klog.V(5).Info("waiting for informer caches to sync")
+	if !cache.WaitForNamedCacheSync("cluster-lifecycle-controller", ctx.Done(), c.clusterSynced) {
+		return
+	}
+
+	wait.UntilWithContext(ctx, c.monitorClusterHealth, ClusterMonitorPeriod)
+}
+
+func (c *Controller) monitorClusterHealth(ctx context.Context) {
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Warningf("failed to list clusters: %v", err)
+		return
+	}
+
+	for i := range clusters {
+		cluster := clusters[i].DeepCopy() // Make a copy from cluster cause we might change it below
+		c.tryUpdateClusterReadyCondition(ctx, cluster, wait.Backoff{
```

Comment:
Since wait.Backoff here is even not passed from upstream, should we treat this as a parameter?